### PR TITLE
DEV: Update Boschloo Exact test

### PR DESCRIPTION
**Pull request by V0lantis**
Update Boschloo Exact test to improve its robustness for small numerical errors

Also, we might want to discuss more about tolerance over `np.isclose`. From [numpy's doc ](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html)

> For finite values, isclose uses the following equation to test whether two floating point values are equivalent.
> 
> absolute(a - b) <= (atol + rtol * absolute(b))
> 
> Unlike the built-in math.isclose, the above equation is not symmetric in a and b – it assumes b is the reference value – so that isclose(a, b) might be different from isclose(b, a).

Since in my case, `b==0`, the absolute tolerance  is `1e-8`, which in my opinion is enough. I didn't test with smaller value though

Also following @mdhaber [comment](https://github.com/scipy/scipy/pull/13951#issuecomment-849213832) :

> The documentation shows a different null hypotheses for each alternative hypothesis. In most hypothesis tests, the null distribution is calculated the same way regardless of the alternative hypothesis, so in those cases I would say that the null hypothesis is the same for all three alternative hypotheses. (Apparently that is not true in the KS test, but I haven't looked into it very closely.) Is it correct here that you really have a different null hypothesis corresponding with each alternative hypothesis?

I am not sure to understand completely your question here. I just learned this way of defining null hypothesis. 
For example : 
Capture d’écran 2021-06-04 à 15 13 44
If we accept **H0**, it means that the alternative is incorrect, and in the above picture, it also means that p_1 \geq p_2 right ?

> I'd still like to figure out why R is making the plot symmetric for the table [[0 2], [6 4]]. It appears that it has taken the left half of your plot and mirrored it to produce the right half. So I asked before: is there a table (e.g. by swapping rows or columns) that would cause your (whole, not just half) plot to be the mirror image of the original? If you put that table into R, what does its plot look like? My guess is that R can somehow tell which half of the range of the nuisance parameter will maximize the p-value and that it is only looking in that half. Does that make sense based on some feature of the table (e.g. that one of the elements is zero)?

I wans't able to dig more into it

### CODE REVIEW

**Comment by mdhaber** [comment]
> Since in my case, b==0, the absolute tolerance is 1e-8, which in my opinion is enough. I didn't test with smaller value though

It is not a matter of opinion, though - unless we're dealing with really extreme tables, there exists a tolerance below which differences between elements of `pvalue` are artifacts of finite precision issues and above which are the differences between elements of `pvalue` are real. `atol=1e-8` probably works for a lot of tables, but the choice of `1e-8` is arbitrary, not yet studied, and I think we can do better. Let me think about this some more.

> I am not sure to understand completely your question here. I just learned this way of defining null hypothesis...

Does gh-12518 help clarify the distinction?

**Comment by mdhaber** [comment]
_I'll do my best to make this make sense and be concise, but it's a bit complicated. 
TLDR: use `np.isclose(..., atol=1e-13)`. We might be able to do something smarter, but this is a better fixed `atol` than the default._

I investigated the effect of the line:
```python3
index_arr = np.logical_or(index_arr, np.isclose(pvalues - fisher_stat, 0))
```
Specifically, I looked for tables for which the tolerance of `np.isclose` might be incorrect. That is, I looked for tables for which some elements of `pvalues` might actually be distinct from `fisher_stat` but would be incorrectly classified as `isclose`/"equal" to `fisher_stat`. This will lead to an incorrect `1` in `index_arr` and presumably some error in the p-value.

To do this, I created two masks:
```
mask1 = np.logical_and(np.isclose(pvalues - fisher_stat, 0), pvalues > fisher_stat)
mask2 = np.logical_and(~np.isclose(pvalues - fisher_stat, 0), pvalues > fisher_stat)
```
`mask1` picks out the elements of `pvalues` that `np.isclose` considers to be "truly" greater than `fisher_stat`. I call these "real" differences.
`mask2` picks out the elements of `pvalues` that `np.isclose` consider to be sufficiently "close" to `fisher_stat` to treat them as "equal", despite the fact that they are numerically a little larger than `fisher_stat`. I call these "false" differences. 

Note that my labels "false" and "real" are as-determined by `np.isclose`. Ultimately, I'm looking for tables for which `np.isclose` will identify "false" differences that are _in reality_ "real" differences. These would lead to an incorrect p-value.

Using these masks, I searched through all 2x2 contingency tables with elements <= 19, looking for "false" differences that _are truly real differences_.  

As each table can have multiple "false" differences and will typically have many "real" differences, I only need to consider the maximum "false" difference and the minimum "real" difference of each table. I filtered out the tables which didn't seem likely to be problematic, and for the remaining tables, I plotted the maximum "false" difference against the minimum "real" difference.

image

You can see that there are many tables for which the maximum "false" difference is very nearly equal in magnitude to the minimum "real" difference (~`1e-8`), so for these tables, I think that the default tolerance of `np.isclose` will make mistakes. We can do better, so I think we should.

You see that there is a big gap in the "false" differences between `1e-14` and `1e-10`. This suggests to me that anything below `1-e14` is truly numerical error and anything above that is truly a different p-value. So one way to mitigate the problem is to choose an `atol` in-between, such as `1e-13`.

As tables get bigger, the smallest true differences (or, more precisely, the maximum incorrectly-labeled "false" differences) tend to decrease:
image

So for sufficiently large tables, a fixed tolerance of `1e-13` will no longer correctly distinguish between false and real differences. But I suspect that will not matter much because as tables get larger, the p-values probably get less sensitive to this sort of error.

**Comment by V0lantis** [comment]
> > I am not sure to understand completely your question here. I just learned this way of defining null hypothesis...
> 
> Does [gh-12518](https://github.com/scipy/scipy/issues/12518) help clarify the distinction?

Ok, I understood better your question. I went through Boschloo's and Fisher's papers, and they both have written : 
Capture d’écran 2021-06-13 à 12 34 24

So I guess, this is the way to go

**Comment by V0lantis** [comment]
I merged master to take into account the modifications of 
Also, I run the benchmarks, and I am very pleased to see a great improvement from the boost hypothests implementaitons : 

```
All benchmarks:

       before           after         ratio
     [eaf53115]       [05574ea4]
     <stats/update_boschloo~4>       <stats/update_boschloo>
         12.8±2ms         8.80±1ms    ~0.69  stats.CorrelationFunctions.time_boschloo_exact('greater')
         12.3±2ms         7.27±2ms    ~0.59  stats.CorrelationFunctions.time_boschloo_exact('less')
         24.3±3ms         16.6±2ms    ~0.68  stats.CorrelationFunctions.time_boschloo_exact('two-sided')

```
Nearly half the time ! 😁

**Review by mdhaber — APPROVED** [review]
Looks good to me. Were you hoping for the modification in this PR to be backported so that it is released with 1.7?

**Comment by V0lantis** [comment]
> Looks good to me. Were you hoping for the modification in this PR to be backported so that it is released with 1.7?

It could have be nice, but otherwise, Boschloo's implementation is still correct anyway and works for most of the case

**Comment by mdhaber** [comment]
If @tupui is able to review in time, I will suggest that this be backported. If not, we will save it for 1.7.1.

**Review by tupui — APPROVED** [review]
Thanks @V0lantis. Changes seem sensible and thanks to @mdhaber's graph the hard coded value seems justified.

**Comment by tupui** [comment]
As the original version is still  correct as you indicated, I would keep this for 1.7.1. IMO we should not delay the release with another RC if just for this. As I understand, there is not much more left before the final version.

**Comment by tupui** [comment]
I should have merged sorry for the delay.
